### PR TITLE
x/oracle: refactor data

### PR DESCRIPTION
**Pull request by fedekunze**
Notable changes:

- custom oracle data handling functions
- UniswapPair now uses sdk.Dec as default
- refactor EndBlock logic for better performance
- aggregated data per block is now set to store
- add telemetry metrics
- validation on msgs, params, and genesis state

### CODE REVIEW

File: proto/oracle/v1/oracle.proto [hunk 1]
```diff
@@ -5,20 +5,27 @@ import "gogoproto/gogo.proto";
 
 option go_package = "github.com/peggyjv/sommelier/x/oracle/types";
 
-
-// UniswapData is an implementation of OracleData
+// UniswapData represents the necessary data for a given uniswap pair.
 message UniswapData {
-	repeated UniswapPair pairs = 1 [
-		(gogoproto.nullable) = false
+	string        id   = 1;
```
Comment (jackzampolin) [inline proto/oracle/v1/oracle.proto:10]: So this is used to also parse the JSON from the graph. This will break the app in a number of places.

Comment (jackzampolin) [reply proto/oracle/v1/oracle.proto:10]: could easily make a to parse tho...
```go
type Uniswap struct {
  Pairs []UniswapPair `json:"data"`
}
```

File: proto/oracle/v1/oracle.proto [hunk 2]
```diff
@@ -5,20 +5,27 @@ import "gogoproto/gogo.proto";
 
 option go_package = "github.com/peggyjv/sommelier/x/oracle/types";
 
-
-// UniswapData is an implementation of OracleData
+// UniswapData represents the necessary data for a given uniswap pair.
 message UniswapData {
```
Comment (jackzampolin) [inline proto/oracle/v1/oracle.proto:9]: This should be `UniswapPair`

File: proto/oracle/v1/oracle.proto [hunk 3]
```diff
@@ -27,13 +34,30 @@ message UniswapPair {
...
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
+    (gogoproto.nullable)   = false
+	];
 }
 
 // UniswapToken is the returned uniswap token representation
 message UniswapToken {
-    string decimals = 1 [(gogoproto.jsontag) = "decimals"];
-    string id       = 2 [(gogoproto.jsontag) = "id"];
+		// TODO: what's the ID?
```
Comment (jackzampolin) [inline proto/oracle/v1/oracle.proto:59]: The ID is the contract address of the ERC20 token.

**Comment by jackzampolin** [comment]
@fedekunze and I discussed offline a bit. We are agreed on direction here but think this is for after initial testnet. Hes working on finishing up https://github.com/PeggyJV/sommelier/pull/14 first.

File: x/oracle/types/oracle.go [hunk 4]
```diff
@@ -124,6 +116,62 @@ func (up UniswapPair) MarshalJSON() ([]byte, error) {
 	return sdk.SortJSON(bz)
 }
 
+// Compare checks that the current pair is within the target range
+func (up UniswapPair) Compare(aggregatedData OracleData, target sdk.Dec) bool {
```
Comment (jackzampolin) [inline x/oracle/types/oracle.go:121]: This is a great start that we can flesh out later.

File: proto/oracle/v1/tx.proto [hunk 5]
```diff
@@ -35,11 +35,8 @@ message MsgOracleDataPrevoteResponse {}
 
 // MsgOracleDataVote - sdk.Msg for submitting arbitrary oracle data that has been prevoted on
 message MsgOracleDataVote {
-    repeated string              salt        = 1;
-    repeated google.protobuf.Any oracle_data = 2 [
-        (cosmos_proto.accepts_interface) = "OracleData"
-    ];
-    string                       signer      = 3;
+    OracleVote vote = 1;
```
Comment (jackzampolin) [inline proto/oracle/v1/tx.proto:38]: Are we pulling the salt out?

Comment (jackzampolin) [reply proto/oracle/v1/tx.proto:38]: I see this in the protofiles now.

File: x/oracle/keeper/abci.go [hunk 6]
```diff
@@ -6,164 +6,197 @@ import (
...
+
+			// add oracle data to maps
+			submittedFeedMap[oracleData.Type()][validatorAddr.String()] = oracleData
+
 		}
 
-		// if the voted_power/total_power > params.VoteThreshold then we store the averages in the store
-		storeAverages := sdk.NewDec(votedPower).Quo(sdk.NewDec(totalPower)).GT(params.VoteThreshold)
+		// delete the vote as it has already been processed
+		// TODO: consider keeping the votes for a few voting windows in order to
```
Comment (jackzampolin) [inline x/oracle/keeper/abci.go:81]: I think saving the votes will be too expensive. cc @zmanian

File: x/oracle/keeper/abci.go [hunk 7]
```diff
@@ -6,164 +6,197 @@ import (
...
+	})
+
+	// iterate over the full list of validators to increment miss counters if they didn't vote
+	totalPower := int64(0)
+	k.stakingKeeper.IterateBondedValidatorsByPower(ctx, func(_ int64, validator stakingtypes.ValidatorI) bool {
+		totalPower += validator.GetConsensusPower()
+		validatorAddr := sdk.AccAddress(validator.GetOperator())
+
+		if !validatorVotesMap[validatorAddr.String()] {
+			// TODO: this is wrong because the feeder could submit a single oracle data uniswap pair instead of all the required
```
Comment (jackzampolin) [inline x/oracle/keeper/abci.go:94]: The way I had this in the prior implemenation is that we required a # of different markets (i.e. 100) from each validator.

File: x/oracle/keeper/abci.go [hunk 8]
```diff
@@ -6,164 +6,197 @@ import (
...
-		for typ, dataCollection := range collectionMap {
-			// first, lets delete the old data
-			k.DeleteOracleData(ctx, typ)
+		return false
+	})
 
-			// then we compute the "average"
-			avg := types.GetAverageFunction(typ)(dataCollection)
+	// if the voted_power/total_power > params.VoteThreshold then we store the averages in the store
+	// TODO: check for sample size too as there could be only a few feeders with a large voting power that submitted the
```
Comment (jackzampolin) [inline x/oracle/keeper/abci.go:104]: This seems like a v2 feature.

File: x/oracle/keeper/abci.go [hunk 9]
```diff
@@ -6,164 +6,197 @@ import (
...
-		// averages to define which are eligable for rewards
-		for dataType, vals := range detailedMap {
-			for val, data := range vals {
-				rewardEligable[val] = false
-				if averageMap[dataType].Valid(data) {
-					rewardEligable[val] = true
-				}
-			}
-		}
+		// TODO: delete the oracle old data
```
Comment (jackzampolin) [inline x/oracle/keeper/abci.go:119]: Why is this still todo? Do we need to decide how many periods of data we are saving?

File: x/oracle/keeper/abci.go [hunk 10]
```diff
@@ -6,164 +6,197 @@ import (
...
+		// k.DeleteOracleData(ctx, dataType, dataType, oracleData)
 
-		// slash validators who have missed to many votes
-		k.IterateMissCounters(ctx, func(val sdk.AccAddress, counter int64) bool {
+		// aggregate the date using the handler set up during app initialization
+		aggregatedData, err := k.oracleHandler(ctx, oracleData)
+		if err != nil {
+			// TODO: ensure correctness or consider logging instead?
```
Comment (jackzampolin) [inline x/oracle/keeper/abci.go:125]: I think we should assume that the data is sufficiently checked on the way in that by the time we get here this should panic if there is an error.

File: x/oracle/keeper/abci.go [hunk 11]
```diff
@@ -6,164 +6,197 @@ import (
...
-			consensusAddr, err := validator.GetConsAddr()
-			if err != nil {
-				// TODO: check why this could error and ensure correctness
-				panic(err)
+	// Compare each validators vote for each data type against the
+	// averages to define which are eligable for rewards
+	for dataType, vals := range detailedMap {
+		for val, data := range vals {
+			rewardEligable[val] = false
+			// TODO: check if the data provided is within x% of the average? I would assume x would have to be low:
```
Comment (jackzampolin) [inline x/oracle/keeper/abci.go:143]: Yeah I think this should be good. I see you added param for this.

File: x/oracle/keeper/abci.go [hunk 12]
```diff
@@ -6,164 +6,197 @@ import (
...
-		})
+		}
 
-		// TODO: reward validators
-		// TODO: Setup module account for oracle module
+		// slash validator below the minimum vote threshold
+		validator := k.stakingKeeper.Validator(ctx, sdk.ValAddress(validatorAddr))
+		if validator == nil {
+			// validator not found, continue with the next counter
+			// TODO: ensure correctness
```
Comment (jackzampolin) [inline x/oracle/keeper/abci.go:164]: Don't understand this comment.

File: x/oracle/keeper/abci.go (hunk 6 above)
Comment (fedekunze) [reply x/oracle/keeper/abci.go:114]: agree, maybe for 20 blocks should be sufficient?

File: cmd/oracle-feeder/cmd/start.go [hunk 13]
```diff
@@ -103,49 +112,65 @@ func (c *Coordinator) handleBlock(blockEvent ctypes.ResultEvent) error {
...
 	if err != nil {
 		return err
 	}
-	c.UD = ud
-	c.Salt = genrandstr(6)
-	c.Hash = oracle.DataHash(c.Salt, ud.CannonicalJSON(), c.Val)
-	return config.BroadcastTx(c.Ctx, c,
-		oracle.NewMsgOracleDataPrevote([][]byte{c.Hash}, c.Addr))
+
+	c.salt = genrandstr(6)
```
Comment (jackzampolin) [inline cmd/oracle-feeder/cmd/start.go:160]: are you still storing the feed on the coordinator here?

**Comment by jackzampolin** [comment]
Ok, so we have pushed through all the keeper errors here and have the data persisting. There is one remaining TODO to query the data now. @fedekunze is working on a separate PR for that. Going to go ahead and merge here. See followup PR for querability.

**Pull request merged**